WIP: Still fleshing out some code samples and framing [draft] (#4376)
viamrobotics/docs: docs/data-ai/capture-data/lorawan.md

---
title: "Create a LoRaWAN network"
linkTitle: "Create a LoRaWAN network"
weight: 60
type: "docs"
description: "Configure a gateway and nodes to communicate over the LoRaWAN protocol."
tags: ["sensor", "components", "lorawan", "gateway", "node"]
icon: true
images: ["/icons/components/sensor.svg"]
---

[LoRaWAN (Long Range Wide Area Network)](https://lora-alliance.org/) enables sensor communication spanning kilometers with minimal power usage.

To collect LoRaWAN sensor data with Viam, use the [`lorawan`](https://app.viam.com/module/viam/lorawan) {{< glossary_tooltip term_id="module" text="module" >}}.

> You can find the module's source code in the [lorawan GitHub repository](https://github.com/viam-modules/lorawan).

Reviewer: ```suggestion
```

Author: Removed!

## Hardware requirements

- A [supported LoRaWAN gateway](https://github.com/viam-modules/lorawan?tab=readme-ov-file#lorawan-gateway-models-provided)
- A [supported LoRaWAN node](https://github.com/viam-modules/lorawan?tab=readme-ov-file#lorawan-sensor-models-provided)

## Architecture

You can use Viam to create LoRaWAN networks containing one gateway and multiple nodes.
In a LoRaWAN network, information flows in two directions:

- **uplinks** transmit from nodes to gateways
- **downlinks** transmit from gateways to nodes

![A LoRaWAN network consisting of a single gateway and multiple nodes](/components/sensor/lorawan.png)

### Nodes

Nodes measure data like any other sensor.
Instead of physically connecting to a machine over GPIO pins or USB, nodes communicate with your machine using the wireless LoRaWAN protocol.
LoRaWAN supports communication over distances of up to 10 kilometers.
Nodes typically run off of battery power; a single coin cell battery can power a node for weeks or months.

### Gateway

Gateways collect data generated by LoRaWAN nodes.

> A gateway is a device that communicates with nodes over LoRaWAN.

Reviewer: ```suggestion
The gateway is the device that communicates with nodes over LoRaWAN and that connects to Viam.
```

Author: Incorporating this feedback in my next commit, thanks.

Add a gateway to a Viam machine to sync data from many sensors to a single collection in the Viam cloud, where you can aggregate and visualize your data.

> With Viam's `lorawan` module, your gateway also acts as a network server.

Author: Agreed, removing this sentence in my next commit.

1. Click **Save** in the top right to apply your changes.

{{% /tab %}}
{{% tab name="JSON Configuration" %}}

In the `components` section of your machine configuration, add the following objects:

```json
{
  "name": "<lorawan-gateway-board>",
  "api": "rdk:component:board",
  "model": "viam:raspberry-pi:<your-pi-model>",
  "attributes": {}
},
{
  "name": "<your-gateway-name>",
  "api": "rdk:component:sensor",
  "model": "viam:lorawan:<gateway-sensor-name>",
  "attributes": {
    "board": "<lorawan-gateway-board>",
    "spi_bus": "<spi-bus-number>",
    "region_code": "<region-code>"
  }
}
```

Choose an appropriate board model from the [board components registry](/operate/reference/components/board/).

Choose an appropriate gateway model from the following options:

- `viam:lorawan:sx1302-waveshare-hat`: Waveshare LoRaWAN SX1302 Gateway HAT
- `viam:lorawan:sx1302-hat-generic`: generic model for all other peripherals built using the SX1302 or SX1303 chips
- `viam:lorawan:rak7391`: RAK7391 WisGate Connect

Configure attributes based on the tables below:

{{% /tab %}}
{{< /tabs >}}

You must configure the following attributes for SX1302 and SX1303-based LoRaWAN gateways:

> <!-- prettier-ignore -->

Author: I eliminated everything but the essentials, but I think this comment convinces me to transmute these into bullet points instead. Tables are a tad too formal/referential for this context.

| `gateways` | string[] | **Required** | Name of the [gateway component](#add-a-gateway) in your Viam configuration. |

You must configure the following attributes for ABP nodes:

<!-- prettier-ignore -->
| Name | Type | Required? | Description |
| ---- | ---- | --------- | ----------- |
| `join_type` | string | Optional | The [activation protocol](#activation-protocols) used to secure this network. <br> Default: "OTAA". <br> Options: "OTAA", "ABP" |
| `dev_eui` | string | **Required** | The **device EUI (Extended Unique Identifier)**, a unique 64-bit identifier for the LoRaWAN device in hexadecimal format (16 characters). Found on your device or in device packaging. |
| `app_key` | string | Optional | The 128-bit hexadecimal AES **application key** used for device authentication and session key derivation. Required for OTAA activation protocol. Found in the device datasheet. <br> On `milesight-ct101` and `milesight-em310-tilt`, defaults to `5572404C696E6B4C6F52613230313823`. |
| `gateways` | string[] | **Required** | Name of the [gateway component](#add-a-gateway) in your Viam configuration. |

> For generic nodes, you must also configure a [`decoder_path`](#decoder-script).

Reviewer: what are generic nodes? 

Also the connection between decoder_path and what it should link to is a bit missing:
```suggestion
For generic nodes, you must also configure a `decoder_path` for the [decoder script](#decoder-script).
```

{{< tabs >}}
{{% tab name="OTAA" %}}

```json
{
  "name": "<your-node-name>",
  "api": "rdk:component:sensor",
  "model": "viam:lorawan:<node-name>",
  "attributes": {
    "dev_eui": <device-eui>,
    "app_key": <application-key>,
    "gateways": [<gateway-name>]
  }
}
```

{{% /tab %}}
{{% tab name="ABP" %}}

```json
{
  "name": "<your-node-name>",
  "api": "rdk:component:sensor",
  "model": "viam:lorawan:<node-name>",
  "attributes": {
    "join_type": "ABP",
    "dev_addr": <device-address>,
    "app_s_key": <application-session-key>,
    "network_s_key": <network-session-key>,
    "gateways": [<gateway-name>]
  }
}
```

{{% /tab %}}
{{< /tabs >}}

Choose an appropriate node model from the following options:

- `viam:lorawan:dragino-LHT65N`: Dragino LHT65N temperature and humidity sensor.
- `viam:lorawan:dragino-WQSLB`: Dragino WQS-LB water quality sensor
- `viam:lorawan:milesight-ct101`: Milesight CT101 current sensor
- `viam:lorawan:milesight-em310-tilt`: Milesight EM310-TILT sensor
- `viam:lorawan:node`: Any LoRaWAN sensor that is:
  - Class A
  - supports either the `US915` or `EU868` frequency band
  - uses LoRaWAN MAC specification version 1.0.3

> Configure attributes based on the tables below:

Author: ```suggestion
Configure attributes based on the descriptions below:
```

For generic peripherals, you must also configure `spi_bus`, `power_en_pin`, and `path`. For a full list of attributes, see the [module README](https://github.com/viam-modules/lorawan?tab=readme-ov-file#configuration-for-lorawan-gateways).

You must configure the following attributes for RAK7391 gateways:

- `board`: The name of the [board component](/operate/reference/components/board/) that represents the Raspberry Pi Compute Module inside the RAK7391. Used for GPIO pin control.

## Add a node

Complete the following steps to configure your node:

{{< tabs >}}
{{% tab name="Builder" %}}

1. Navigate to the **CONFIGURE** tab of your machine's page in the [Viam app](https://app.viam.com).
1. Click the **+** icon next to your machine part in the left-hand menu and select **Component or service**.
1. Select the `sensor` type, type `lorawan`, then select the `lorawan` model that matches the name of your node.
   If the name of your node does not appear in the list, choose the generic `lorawan:node` option.
1. Click **Add module**, and enter a name for your node.
1. Click **Create** to add the model and module to your machine.

> 1. Configure the model attributes as follows, according to the node type:
> 1. Click **Save** in the top right to apply your changes.

Reviewer 2: Sorry I may have made a mistake here in my suggestion in the last review.
```suggestion
1. Configure the model attributes as follows, according to the node type:
```
or something like this, or move "Click save..." to the end
```suggestion
1. Configure the model attributes as described below, according to the node type.
1. Click **Save** in the top right to apply your changes.
```

{{% /tab %}}
{{< /tabs >}}

{{% /tab %}}
{{< /tabs >}}

### Send a downlink message to a node

{{< tabs >}}
{{% tab name="Viam web app" %}}

> To restart a node from the Viam web app, specify a hexadecimal string in the following DoCommand input in the **CONTROL** tab of your machine page:

Reviewer: ```suggestion
To restart a node from the Viam web UI, specify a hexadecimal string in the following DoCommand input in the **CONTROL** tab of your machine page:
```

```typescript
const node = new SensorClient(client, "<your_node_name>");

await node.doCommand({ <command>: <value> });
```

{{% /tab %}}
{{< /tabs >}}

{{% /tab %}}
{{< /tabs >}}

The Dragino WQS-LB supports the following calibration commands:

- `calibrate_ph`
- `calibrate_ec`
- `calibrate_t`
- `calibrate_orp`

Use the calibration processes below to calibrate your sensor with these commands:

> #### Calibrate the pH probe

Reviewer: Is this a common enough thing that we want to cover it here? Leaving that up to you but it feels odd to me

Author: It is a requirement to properly use this sensor model, which is one of our small number of officially supported sensors. And the pattern is generalizable across all other sensors that require calibration. So I'm leaning 'include' unless someone feels really strongly otherwise. I understand how it might feel odd but I think it's important to draw attention to this so folks don't wind up with a lot of uncalibrated sensors with unreliable readings.
